pbj-core/pbj-runtime/src/main/java/com/hedera/hashgraph/pbj/runtime/io/ReadOnlyDataBuffer.java

```diff
@@ -0,0 +1,230 @@
+package com.hedera.hashgraph.pbj.runtime.io;
+
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * A Buffer backed by a ByteBuffer that implements {@code DataInput} and {@code DataOutput}.
+ */
+public sealed class ReadOnlyDataBuffer implements DataInput permits DataBuffer, ReadOnlyOffHeapDataBuffer {
+
+    /** ByteBuffer used as backing buffer for this DataBuffer */
+    protected ByteBuffer buffer;
```

Comment:
Should this be final?

```suggestion
    protected final ByteBuffer buffer;
```